### Project : Exploring Ebay Car Sales Data

**Project Description**
The project clean and analyze a sample dataset from the Classied Ad section of a German Ebay website, which was originally scraped and uploaded to Kaggle by user **orgesleka**.
Original dataset is no longer available on Kaggle but can still be accessed [here](https://data.world/data-society/used-cars-data)

The main tools in this project will be pandas and numpy.

In [2]:
import numpy as np
import pandas as pd
import csv


In [110]:
# upload the dataset into pandas 
""" Note unicode UTF-8 is not working for this dataset """
auto = pd.read_csv('autos.csv', encoding='Latin-1')

In [113]:
# make a copy of the dataset so that we can keep the origin data intact  
auto_copy = auto.copy()

In [114]:
auto_copy.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [115]:
auto_copy.tail()

dateCrawled                                               name  \
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

      brand notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
49995  audi              nein  2016-03-27 00:00:00             0       82131   
49996  opel              nein  2016-03-28 00:00:00             0       44807   
49997  fiat              nein  2016-04-02 00:00:00             0       73430   
49998  audi              nein  2016-03-08 00:00:00             0       35683   
49999  opel              nein  2016-03-13 00:00:00             0       45897   

                  lastSeen  
49995  2016-04-01 13:47:40  
49996  2016-04-02 14:18:02  
49997  2016-04-04 11:47:27  
49998  2016-04-05 16:45:07  
49999  2016-04-06 21:18:48

In [116]:
# check dataset info
auto_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## Clean Columns

In [117]:
# printing column list
col_name_origin = auto_copy.columns

In [118]:
col_name_origin

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [122]:
corrected_col_names = {'dateCrawled':'date_crawled',
       'name':'name', 'seller':'seller', 'offerType':'offer_type', 
       'price':'price', 'abtest':'abtest',
       'vehicleType':'vehicle_type', 'yearOfRegistration': 'registration_year',
       'gearbox':'gearbox', 'powerPS':'power_ps', 'model':'model',
       'odometer':'odometer', 'monthOfRegistration':'registration_month', 
       'fuelType':'fuel_type', 'brand':'brand',
       'notRepairedDamage':'not_repaired_damage', 'dateCreated':'ad_created', 
       'nrOfPictures':'nr_of_pictures', 'postalCode':'postal_code',
       'lastSeen':'last_seen'}

In [123]:
# replace original column list with corrected_col_names list
auto_copy.rename(corrected_col_names, axis=1, inplace = True)

In [124]:
# print corrected column list
auto_copy.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'not_repaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

**Note** We changed the original column names which are in Camelcase into Snakecase to make it easier to read and process; for example, 'notRepaireDamage' was changed into 'not_repaired_damage'. We also changed the following columns names into new names: 

    - yearOfRegistration to registration_year
    - monthOfRegistration to registration_month
    - notRepairedDamage to unrepaired_damage
    - dateCreated to ad_created



## Initial Statistic Exploration

In [125]:
# Look at discriptive stats
auto_copy.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       not_repaired_damage           ad_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### It seems that couples text column, unique numbers only 2, which mean almost all the values are the same
    - seller
    - offerType
And the nr_of_pictures and registration_year looks odd, we need to further investigate.

In [126]:
# look at nr_of_picture colum 
auto_copy['nr_of_pictures'].describe()



count    50000.0
mean         0.0
std          0.0
min          0.0
25%          0.0
50%          0.0
75%          0.0
max          0.0
Name: nr_of_pictures, dtype: float64

In [127]:
# check seller column
auto_copy['seller'].describe()

count      50000
unique         2
top       privat
freq       49999
Name: seller, dtype: object

In [129]:
# check offerType col
auto_copy['offer_type'].describe()

count       50000
unique          2
top       Angebot
freq        49999
Name: offer_type, dtype: object

**As we can see that all values of nr_of_pictures are zeros, so we can drop the column as well as 'seller' and 'offerType' columns.**

In [132]:
# dropping seller, offerType and nr_of_picture columns
auto_copy = auto_copy.drop(['seller', 'offer_type', 'nr_of_pictures'], axis = 1)

In [133]:
# check to make sure the columns dropped
auto_copy.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'not_repaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

### **Checking *price* and *odometer* columns**

In [149]:
# price column
print(auto_copy['price'].head())
print(auto_copy['price'].describe())

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object
count     50000
unique     2357
top          $0
freq       1421
Name: price, dtype: object


In [146]:
# convert into str first
auto_copy['price'] = auto_copy['price'].astype(str)

In [151]:
# clean the price columns and convert into int

auto_copy['price'] = (auto_copy['price'].str.replace('$','')
                  .str.replace(',','')
                  .astype(int)
                 )

In [153]:
# Explore Price column's stats
print(auto_copy['price'].value_counts().head())
auto_copy['price'].describe()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

**Notice that there are 1421 car sold for $0, which is about 2% of cars sold.
    Also, the mean price is around 9840 and max price is 10 million. 
    Let's further explore the highest price.**

In [155]:
print(auto_copy['price'].value_counts().sort_index(ascending = False).head(20))
print(auto_copy['price'].value_counts().sort_index(ascending = True).head(20))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


**There a few of listing with price below \$30 including 1421 listed for \$0. 
Since this site is aution, we can conclude that the reasonable prices fall between \$1 and \$350000 
So, we will remove any listings with price outside this range.**

In [157]:
auto_copy = auto_copy[auto_copy['price'].between(1,350000)]
auto_copy['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

### Explore Odometer column

In [158]:
auto_copy['odometer'].value_counts()

150,000km    31414
125,000km     5057
100,000km     2115
90,000km      1734
80,000km      1415
70,000km      1217
60,000km      1155
50,000km      1012
5,000km        836
40,000km       815
30,000km       780
20,000km       762
10,000km       253
Name: odometer, dtype: int64

In [161]:
# clean odometer column and convert it into int
auto_copy['odometer'] = (autos['odometer'].replace('km', '')
                    .replace(',','')
                    .astype(int)
                    )
auto_copy['odometer'].value_counts()

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer, dtype: int64

### Explore date column

In [162]:
date_cols = auto_copy[['date_crawled', 'last_seen', 'ad_created', 'registration_month', 'registration_year']]
date_cols.head()

date_crawled            last_seen           ad_created  \
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00   
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00   
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00   
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00   
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

**We can see that the first 3 columns (date_crawled, last_seen, ad_created) are ind date format. And the first 10 characters is date.
**

### Calculate distribuition of dates in the 3 columns

In [170]:
# date_crawled column
(auto_copy['date_crawled'].str[:10]
 .value_counts(normalize = True, dropna=False)
 .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

**Look like the listing was crawled over a period of about a month and the distribution is rather uniform.**

In [169]:
# Calculate distribution of ad_created
(auto_copy['ad_created'].str[:10]
 .value_counts(normalize = True, dropna=False)
 .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

**It seems like the majority of the ads created dates fall within 1-2 months of listing dates but there are a few of them were as old as 9 months.**

In [168]:
# Calculate distribution of last_seen column
(auto_copy['last_seen'].str[:10]
.value_counts(normalize=True, dropna=False)
.sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

**The crawler recorded the date it last saw any listing, which allows us to see on what day a listing was removed, presumably because the car was sold.**

**The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values have to do with the crawling period ending and don't indicate car sales.**

### Let's look at `registration_year` column

In [171]:
auto_copy['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

**Notice that the min year (1000 ) and max (year 9999 ) registration_year don't seem to make sense as car could not be registered 1000 years ago, way before the car was invented, the max year (9999) is also way out of reality. So, any cars with registration year above 2016 is inaccurate. Now let's find the car that fall between the year 1900 and 2016.**

In [173]:
max_accept_yr = 2016
min_accept_yr = 1900

In [192]:
((~auto_copy['registration_year'].between(min_accept_yr, max_accept_yr))
 .sum() / auto_copy.shape[0])


0.038793369710697

**Since the values outside of acceptable registration year range,
we can drop these rows**

In [195]:
auto_copy = auto_copy[auto_copy['registration_year'].between(min_accept_yr, max_accept_yr)]
auto_copy['registration_year'].value_counts(normalize=True).sort_index(ascending=False).head(10)

2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
2011    0.034768
2010    0.034040
2009    0.044665
2008    0.047450
2007    0.048778
Name: registration_year, dtype: float64

### Explore price by brand

In [260]:
# top 20 brands
top_brands = auto_copy['brand'].value_counts(normalize=True)

# brands with percentage greater than 5%
top6_brands = top_brands[top_brands >= 0.05].index
top6_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

### Aggregate the top 6 brands with percentage greater than 5% in top_6 dict with brand names as keys and mean as values

In [299]:
# initialize empty dict
top6_brand_mean_price = {}

for brand in top6_brands:
    #subset auto_copy df by each brand
    brand_df = auto_copy[auto_copy['brand'] == brand]
    mean_price = brand_df['price'].mean()
    # assign brand as key and mean_price as value to top6_brand_mean_price dict
    top6_brand_mean_price[brand] = int(mean_price)

top6_brand_mean_price


audi             9336
bmw              8332
ford             3749
mercedes_benz    8628
opel             2975
volkswagen       5402
Name: mean_price, dtype: int64

In [264]:
auto_copy.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'not_repaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

### Explore Mileage

In [324]:
# create empty mean_mileage dict
mean_mileage = {}

# filter out average mileage for each brand
for brand in top6_brands:
    # subset df for each brand
    brand_df = auto_copy[auto_copy['brand'] == brand]
    # mean_mileage
    mean_mile = brand_df['odometer'].mean()
    mean_mileage[brand] = int(mean_mile)
    
# convert mean_mileage dict in pandas Series
mean_price_serie = pd.Series(top6_brand_mean_price, name = 'mean_price')
mean_mileage_serie = pd.Series(mean_mileage, name='mean_mileage')

# convert the two series into df
df = pd.DataFrame(mean_mileage_serie)
df['mean_price'] = mean_price_serie

# alternative way to combine series into data frame
df1 = pd.concat([mean_mileage_serie, mean_price_serie], axis=1)
df

mean_mileage  mean_price
audi                 129157        9336
bmw                  132572        8332
ford                 124266        3749
mercedes_benz        130788        8628
opel                 129310        2975
volkswagen           128707        5402

In [318]:
df1

mean_mileage  mean_price
audi                 129157        9336
bmw                  132572        8332
ford                 124266        3749
mercedes_benz        130788        8628
opel                 129310        2975
volkswagen           128707        5402